crates/ra_hir/src/ty.rs

```diff
@@ -872,6 +873,155 @@ impl<'a, D: HirDatabase> InferenceContext<'a, D> {
+        let body = Arc::clone(&self.body); // avoid borrow checker problem
+
+        let ty = match &body[pat] {
+            Pat::Tuple(ref args) => {
+                // this can probably be done without cloning/ collecting
+                let expectations = match expected.ty {
+                    Ty::Tuple(ref tuple_args) if args.len() == tuple_args.len() => {
+                        tuple_args.iter().cloned().collect()
+                    }
+                    _ => vec![Ty::Unknown; args.len()],
+                };
+
+                let inner_tys = args
+                    .iter()
+                    .zip(expectations.into_iter())
+                    .map(|(&pat, ty)| self.infer_pat(pat, &Expectation::has_type(ty)))
+                    .collect::<Vec<_>>()
+                    .into();
+
+                Ty::Tuple(inner_tys)
+            }
+            Pat::Ref { pat, mutability } => {
+                let expectation = match expected.ty {
+                    Ty::Ref(ref sub_ty, exp_mut) if *mutability == exp_mut => {
+                        Expectation::has_type((&**sub_ty).clone())
+                    }
+                    _ => Expectation::none(),
+                };
+                let subty = self.infer_pat(*pat, &expectation);
+                Ty::Ref(subty.into(), *mutability)
+            }
+            Pat::TupleStruct {
+                path: ref p,
+                args: ref subpats,
+            } => self.infer_tuple_struct(p.as_ref(), subpats),
+            Pat::Struct {
+                path: ref p,
+                args: ref fields,
+            } => self.infer_struct(p.as_ref(), fields),
+            Pat::Path(path) => {
+                // is this right?
+                self.module
+                    .resolve_path(self.db, &path)
+                    .take_values()
+                    .map_or(Ty::Unknown, |resolved| self.db.type_for_def(resolved))
+            }
+            Pat::Bind {
+                mode,
+                name: _name,
+                subpat,
+            } => {
+                let subty = if let Some(subpat) = subpat {
+                    self.infer_pat(*subpat, expected)
+                } else {
+                    let ty = self.new_type_var();
+                    self.unify(&ty, &expected.ty);
+                    let ty = self.resolve_ty_as_possible(ty);
+                    ty
+                };
+
+                match mode {
+                    BindingAnnotation::Ref => Ty::Ref(subty.into(), Mutability::Shared),
+                    BindingAnnotation::RefMut => Ty::Ref(subty.into(), Mutability::Mut),
+                    BindingAnnotation::Mutable | BindingAnnotation::Unannotated => subty,
+                }
+            }
+            _ => Ty::Unknown,
+        };
+        // use a new type variable if we got Ty::Unknown here
+        let ty = self.insert_type_vars_shallow(ty);
+        self.unify(&ty, &expected.ty);
```

Comment:
I don't think that's actually correct here, for a `ref x` pattern you'd try to unify `&T` with `T`. do we actually need to unify anything here?

Reply:
Yeah, I now see that you are totally right. I was creating a new variable and immediately unifying that with the expected type :man_facepalming: Now just returning the expected type :-)